File: editorial/chap5-6/prob5-6-2.cpp
#include <iostream>
using namespace std;

const long long mod = 1000000007;
long long N;
long long A[300009];
long long power[300009];

int main() {
	// 入力
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> A[i];
	
	// 2^i を求める
	power[0] = 1;
	for (int i = 1; i <= N; i++) {
		power[i] = (2 * power[i - 1]) % mod;
	}
	
	// 答えを求める
	long long Answer = 0;
	for (int i = 1; i <= N; i++) {
		Answer += power[i - 1] * A[i];
		Answer %= mod;
	}
	
	// 出力
	cout << Answer << endl;
	return 0;
}

File: editorial/chap5-6/prob5-6-2.py
# 入力
N = int(input())
A = list(map(int, input().split()))

# 2^i を求める
mod = 1000000007
power = [0 for i in range(N)]
power[0] = 1
for i in range(1, N):
	power[i] = (2 * power[i - 1]) % mod

# 答えを求める
Answer = 0
for i in range(N):
	Answer += power[i] * A[i]
	Answer %= mod

# 出力
print(Answer)